server/src/main/java/org/candlepin/audit/EventBuilder.java

```diff
@@ -40,47 +40,43 @@ public EventBuilder(EventFactory factory, Target target, Type type) {
+        and it will also be used in the future instead of setNewEntity once newEntity is removed.
+        This is part of simplifying the data Events hold.
+     */
+    public EventBuilder setEventData(Eventful entity) {
+        if (entity != null) {
+            // Be careful to check for null before setting so we don't overwrite anything useful
+            if (entity instanceof Named && ((Named) entity).getName() != null) {
+                event.setTargetName(((Named) entity).getName());
             }
-        }
-        if (entity instanceof Entitlement) {
-            event.setReferenceType(Event.ReferenceType.POOL);
-            Pool referencedPool = ((Entitlement) entity).getPool();
-            if (referencedPool != null && referencedPool.getId() != null) {
-                event.setReferenceId(referencedPool.getId());
+            if (entity instanceof Owned) {
+                Owner entityOwner = ((Owned) entity).getOwner();
+                if (entityOwner != null && entityOwner.getId() != null) {
+                    event.setOwnerId(entityOwner.getId());
+                }
             }
-        }
-        if ((String) entity.getId() != null) {
-            event.setEntityId((String) entity.getId());
-            if (entity instanceof ConsumerProperty) {
-                Consumer owningConsumer = ((ConsumerProperty) entity).getConsumer();
-                if (owningConsumer != null && owningConsumer.getId() != null) {
-                    event.setConsumerId(owningConsumer.getId());
+            if (entity instanceof Entitlement) {
+                event.setReferenceType(Event.ReferenceType.POOL);
+                Pool referencedPool = ((Entitlement) entity).getPool();
+                if (referencedPool != null && referencedPool.getId() != null) {
+                    event.setReferenceId(referencedPool.getId());
                 }
             }
-        }
-    }
-
-    public EventBuilder setOldEntity(Eventful old) {
-        // Allow null, but don't do anything
-        if (old != null) {
-            // If the value is non-null and a value is set, we shouldn't allow it to continue
-            if (event.getType() == Type.CREATED) {
-                throw new IllegalArgumentException("You cannot set the old entity for a creation event");
+            if ((String) entity.getId() != null) {
+                event.setEntityId((String) entity.getId());
+                if (entity instanceof ConsumerProperty) {
+                    Consumer owningConsumer = ((ConsumerProperty) entity).getConsumer();
+                    if (owningConsumer != null && owningConsumer.getId() != null) {
+                        event.setConsumerId(owningConsumer.getId());
+                    }
+                }
             }
-            setEventData(old);
-            event.setOldEntity(factory.entityToJson(old));
+            factory.entityToJson(entity);
```

Comment:
@mstead Adding this line actually fixed the issue. setOldEntity() used to serialize the entity in order to set it on the oldEntity field, but actually the serialization serves some other additional purpose here (even thought we don't set the produced json string to anything!). Not sure exactly how this magic works, but I got the hint from this [comment](https://github.com/candlepin/candlepin/pull/1747/commits/3bb339522592e958f80f8f8bdac9e881c111a5b3#diff-4cbebfb7e395944b0d84bfc0a4a42cf5L1054).

Reply:
```java
    protected String entityToJson(Object entity) {
        String newEntityJson = "";
        // TODO: Throw an auditing exception here

        // Drop data on consumer we do not want serialized, Jackson doesn't
        // seem to care about XmlTransient annotations when used here:

        try {
            newEntityJson = mapper.writeValueAsString(entity);
        }
        catch (Exception e) {
            log.warn("Unable to jsonify: " + entity);
            log.error("jsonification failed!", e);
        }
        return newEntityJson;
    }
```

is what should be getting run by `entityToJson`.  I don't see why any side effects would happen in that code.  😕

Reply:
I know. It doesn't make sense. But it works. -_-

Reply:
What was the exact test that was failing before this change and passes now?  I don't like merging code that is magical in how it works.

Reply:
4 test cases in hypervisor_check_in_spec.rb:
- should add consumer to unchanged when comparing empty guest id lists - async
- should add consumer to unchanged when same guest ids are sent - async
- should allow virt-who to update mappings - async
and one more

Reply:
We need to know why the forcing of the serialization fixes these issues. Is the entity later being detached from hibernate & then accessed?

Reply:
This serialization happens during a HypervisorUpdateJob transaction.
The error happens during a GET request that follows, to check the job's result data (and a consumer object is part of that data). So these are definitely two different hibernate sessions, and the customer is detached. Not sure yet why the serialization helps.

Reply:
Apparently, what happens is:
1) We do an asynchronous updateHypervisor request.
2) An UpdateHypervisorJob runs to do the task. Part of the result data of this job is a Consumer object.
3) During this task there is a specific case where sometimes there is nothing to update. If this is the case, the lazy-loaded collections on the Consumer never get loaded.
4) When the job finishes, it saves the result data in binary form in the cp_job.result_data table field (part of which is the Consumer object, without having its collections initialized).
5) When we do a GET /jobs/#{job_id}?result_data=true, the data (and the Consumer object) get fetched, deserialized from binary to Objects and before returning them their serialization to json is attempted by invoking the getters. The Consumer is detached at this point, so an exception is thrown. 

The problem did not exist before because during step 3, the whole object was serialized as 'oldEntity', forcing the collections to get initialized.

As a solution, I used Hibernate.initialize() to initialize the collections during step 3, before serializing into binary and saving to the db.